### ISSUE TITLE
Fix WebSocket handshake HTTP response

### ISSUE THREAD

**Issue by s-kostyuk** [aio-libs/aiohttp#2973]
## Long story short

WebSocket servers must to send an HTTP response with status-code "101 Switching Protocols" in response to the valid "Connection: Upgrade" request according to the RFC 6455, section [1.2](https://tools.ietf.org/html/rfc6455#section-1.2). And only after that it both parties are able to start a communication.

For now, a correct response is prepared in WebSocketResponse but isn't sent. As result, some client applications like WebSockets Android app (links below) just refuse to connect.

## Expected behaviour

aiohttp sends an "101 Switching Protocols" HTTP response immediately after a corresponding HTTP request was received

## Actual behaviour

aiohttp just starts a communication without sending HTTP response to the client. An actual response is (?) sent only when WS connection is closed.

## Steps to reproduce

1. Open Wireshark or any other tool to monitor traffic
2. Create and run a simple echo server as described in docs (https://docs.aiohttp.org/en/stable/web_quickstart.html#websockets)
3. Try to initialize a connection with mentioned client apps (links below)
4. Compare with behavior of other servers (ws://echo.websocket.org or a sample from https://websockets.readthedocs.io/en/stable/intro.html#basic-example)

## Your environment

Python: 3.5.2 
aiohttp server: 2.3.7
OS: elementary OS 0.4.1 Loki
Web Client App: http://websocket.org/echo.html
WebSockets Android app: https://play.google.com/store/apps/details?id=com.quaksire.websocket

**Comment by asvetlov** [comment]
Oops.
Maybe it is related to #2961
I was pretty sure that we do right things but looks like not.
The solution could look like:
```
async def _pre_start(self, request):
        headers, protocol, compress, notakeover = self._handshake(
            request)

    resp = web.Response(status=101, headers=header)
    await resp.prepare(request)
    await resp.write_eof()

    self._reset_heartbeat()
    self._compress = compress
    writer = WebSocketWriter(request._protocol,
                                 request._protocol.transport,
                                 compress=compress,
                                 notakeover=notakeover)
```
Would you make a PR with fix?

**Comment by s-kostyuk** [comment]
> Maybe it is related to #2961

Yes, it's very likely because just after the first received message servers responds with 101 as HTTP Continuation: [image for proof](https://user-images.githubusercontent.com/5122578/39545507-0e3a2c3a-4e5a-11e8-98f1-2c482be2a511.png).

> Would you make a PR with fix?

I can try. But it'll be hard to test the fix on my machine: the latest version of aiohttp requests Python > 3.5.3 which is not (yet ?) provided by mainline repositories for my current distribution

UPD: Found a third-party repo. Maybe it'll work out

**Comment by asvetlov** [comment]
Tools like https://github.com/pyenv/pyenv can install any python version easily.

**Comment by s-kostyuk** [comment]
Thank you

Still investigating. The good news are that I made a mistake.

HTTP response IS actually sent by aiohttp (no modifications made, both 2.3.7 and 3.2.0a0 versions). But for some reason it's not parsed by Wireshark properly and is displayed as "Reassembled PDU" ([image for proof](https://user-images.githubusercontent.com/5122578/39567004-329b9728-4ec6-11e8-82b0-81782263a65a.png)).

So, looks like the handshake logic is valid. But what than caused #2961, an issue with Android WebSockets client application and some strange display in Wireshark? I don't understand at all. Some pause in transmission, maybe?

P.S. It's also needed to check if Content-Type and Transfer-Encoding headers are allowed for WS handshake responses.

**Comment by asvetlov** [comment]
Maybe Android doesn't like chunked encoding?

**Comment by hubo1016** [comment]
Maybe It should not have content-type and transfer-encoding, because 1xx responses do not have content.

**Comment by asvetlov** [comment]
I cannot test it.
Maybe @s-kostyuk could help?

**Comment by s-kostyuk** [comment]
I'll try to disable content-type and transfer-encoding headers later today (EET). For now I'm reading original RFC 6455 to understand how it must to work

**Comment by s-kostyuk** [comment]
Determined the cause for connection issue with "WebSockets" Android app. **In short: it was not an aiohttp issue. Other clients (okhttp, web clients) work fine**

The WebSockets implementation used in "WebSockets" Android app (and probably only in this app) doesn't conform [RFC 7230](https://stackoverflow.com/questions/5258977/are-http-headers-case-sensitive). This application was searching for a `Sec-WebSocket-Accept` header, case sensitively and thus was unable to find a correct `Sec-Websocket-Accept` (note the small `s` letter), sent by aiohttp.

Considering that there is a client-side issue and that there is a client which implemented the protocol improperly, the original issue is not an issue of aiohttp. But if anybody is curious about the fix, here is a small hack for this: https://github.com/s-kostyuk/aiohttp/commit/3fe2336aca0004f3564ca7c3731fea04926f8177 (available also as an [image](https://user-images.githubusercontent.com/5122578/39672055-c51a9ae2-512b-11e8-848c-6c7f1503ce32.png) for history purposes).

And, such library as okhttp Android library was tested with aiohttp server, had successfully connected and is correctly working with aiohttp server.

This issue will be closed to the end of this day if no new information will be found. And, most likely, if something else will be eventually found, then a new issue will be created.

Thanks to everyone

**Comment by asvetlov** [comment]
As workaround we can store websocket related headers exactly in the same case as written in RFC.
Would you make a PR for it?
The change cannot fix everything, the current implementation is correct. But why not?

**Comment by s-kostyuk** [comment]
It can be done but I see some limitations:
1) as mentioned in [Multidict](https://multidict.readthedocs.io/en/stable/multidict.html) documentation, it's better to store CIMultidict Keys as `istr` instances to prevent unwanted string title-casing on each key access attempts;
2) if so, we either got a title-cased string (as now) in WS handshake response or an impart for aiohttp performance (maybe, negligible, but anyway).

And after all, I don't think that this workaround is needed for now. There is only one known client application that is affected by this issue and it's not an aiohttp issue at all.

But anyway, a commit was prepared, tests are passing and, if needed, I can add a change to the CHANGES folder and make a pull request.

**Comment by asvetlov** [comment]
Multidict preserves passed casing and use title-cased version for comparison only.

**Comment by asvetlov** [comment]
You can create a PR with a fix or close the issue -- the decision is on you, I'm ok with both variants.

**Comment by s-kostyuk** [comment]
I'll better close this issue until some other people will be affected

Thank you
